### DISCUSSION TITLE
How to Bulk Delete Images Based on Name Pattern in Immich?

### DISCUSSION THREAD

**Discussion by coltenkrauter** [immich-app/immich#14674] [Q&A] [answered]
I’m in a bit of a bind and could use some guidance on cleaning up my Immich server.

I need to delete approximately **500,000 images** from my Immich server that follow the naming pattern `*-thumbnail.webp`. These files were autogenerated thumbnails at some point, but I accidentally re-uploaded them as part of my library due to a deployment mishap.

### Backstory

- I originally deployed Immich using the TrueCharts app on TrueNAS.
- Due to TrueNAS Scale moving away from TrueCharts ([TrueNAS migration discussion](https://forums.truenas.com/t/migrating-away-from-truecharts-on-truenas-scale/21804)), I set up a new Immich instance using the TrueNAS Scale app.
- During the migration, I pointed the Immich CLI to my photos directory, which also contained the generated thumbnails. As a result, both my library and the thumbnails were uploaded, leading to the month of May 2024 containing a low-resolution copy of my entire library. 
- Later, I switched to Docker Compose for a more stable setup. This time, I was more careful when uploading to not include the thumbnails dir... but I had not yet noticed that I already had thousands of thumbnail images mixed into my library.

### Issue

I would like to delete all `*-thumbnail.webp` files, but I want to ensure that I clear both the files and any associated thumbnails and entries in the Immich database to avoid orphaned references or issues.

I understand that I could theoretically run something like this,

```bash
find /path/to/root -type f -name '*-thumbnail.webp' -delete
```

However, I suspect this wouldn’t clean the Immich database properly, and I'm not sure if there would be a bunch of thumbnails of thumbnails wasting space on my server.

### Question

Is there a recommended or safe way to bulk delete these files and also update the Immich database?  
Or am I better off taking the "nuclear option" of setting up a fresh instance and re-uploading only the correct files?  

The latter would be time-consuming as my library is nearly 5TB in size and would take many days to re-upload, so I’m hoping there’s a more efficient solution.  

Any advice or guidance would be greatly appreciated!

**Comment by bo0tzz** [comment, marked as answer]
From the advanced search menu, you can filter by (part of) filename:

![image](https://github.com/user-attachments/assets/97881549-1ebd-4589-9f07-7958f1c5da66)

Then you could try just select-all and delete, though for half a million assets the UI might not be all too happy about that. A better chance would be to write a little script using that search query against the API directly.

**Reply by coltenkrauter** [reply]
> From the advanced search menu, you can filter by (part of) filename:
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/97881549-1ebd-4589-9f07-7958f1c5da66)
> 
> 
> 
> Then you could try just select-all and delete, though for half a million assets the UI might not be all too happy about that. A better chance would be to write a little script using that search query against the API directly.

Great idea, the API. 🤦
That's what I'll do! Thank you!

I had seen the UI option, but I was concerned that it might not be able to handle the scale.

**Reply by coltenkrauter** [reply]
I whipped up [this script](https://github.com/krauters/immich-utils/blob/main/src/bulk-delete-assets-by-suffix.ts) to bulk delete for me.

**Reply by coltenkrauter** [reply]
Worked great... I cleared cache in my browser... Now, the timeline seems to have some gray placeholder box for files that got deleted...

Just checked the logs, it is showing many of these,

```
Queueing 1000 assets for deletion from the trash
```

So maybe deletions are still in progress?

**Reply by coltenkrauter** [reply]
Okay, found out that I also need to blow away all of the `*-preview.webp` files.... So, that will probably fix me up.

**Reply by coltenkrauter** [reply]
Yup... Things are good now.

**Comment by coltenkrauter** [comment]
I will note that I experienced something like this while all of these thumbnails were uploaded on the same day: #13319
